### ISSUE TITLE
Conditional type returns false positives

### ISSUE THREAD

**Issue by karol-majewski** [microsoft/TypeScript#34397]
**TypeScript Version:**  3.7-Beta

**Search Terms:** infer, conditional types, assignability

**Code**

```ts
type Either<E, T> = E | { value: T };

/**
 * Good. `Either<Error, Response>` is not a subtype of `Either<string, Response>`.
 */
type Nope =
  Either<Error, Response> extends Either<string, Response>
    ? true
    : false;

/**
 * However, if we perform the same check inside a conditional type, it appears to be.
 */
type ResultType<T> =
  T extends Either<string, infer U>
    ? U
    : never;

/**
 * Expected: `never`, received: `Response`.
 */
type T1 = ResultType<Either<Error, Response>>;
```

**Expected behavior:**

`T1` should be `never`.

**Actual behavior:**

`T1` is of type `Response`.

**Playground Link:** 

[Playground Link](https://www.typescriptlang.org/play/?ts=3.7-Beta&ssl=1&ssc=1&pln=20&pc=44#code/C4TwDgpgBAoglsAFhATgHhgGigFQHxQC8sUAPlAN5QBuAhgDYCuEAXLlAL4DcAUDwPQAqQTyiCoAcQD2UgCYA6KAAN4SVBhQopKbACUIAZzBSAdgYh4lUOAagmpwKLSgHGAI1CQoUgGbLVyOgGwChwJgDmeobGZhZK8qKC-Dye0AByUl6EorAIgRpaOlD6RqbmBBAAHsAQJrK2AerBoRFRpbF4OVBQAPxQIcxdUGw+DOa8AsKJUAASUgDuENSo2HB+i1CQKD7aALb9yC60u9AAxsinANbWZnCy0M6nprIIcKYM-eAQq460YJC0FC2YBSKBuCAJMTJVLFQyMejAHBfND4Ig5HBQKo1OoNPJNEJhSI3HyoKAAVU63W6fTJQzYJiWqAmQhEYlglUgpxqsjYSgZyxQSmwKAgpwgcGWPOUJRi5niiWhX1wAEYiLDXAikZAMHj0DBNNo2rKLHguEA)

**Comment by jack-williams** [comment]
I think this is working correctly.

The first one non-distributive and the conditional type resolves to `never` because
`Error | { value: Response }` is not assignable to `string | { value: Response }` as `Error` is not assignable to `string | { value: Response }`.

The latter is distributive, so you effectively have:
```
ResultType<Either<Error, Response>> =
(Error extends string | { value: Response } ? Response : never) |
  ({ value: Response } extends string | { value: Response } ? Response : never)

===> never | Response ===> Response
```

**Comment by karol-majewski** [comment]
Ouch, distributive conditional types strike again! Thank you for your response @jack-williams. I will put some extra braces here and there.
